api/server.go

```diff
@@ -0,0 +1,97 @@
+// Package server provides a server object that represents the GoCert backend
+package server
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/canonical/gocert/internal/certdb"
+	"gopkg.in/yaml.v3"
+)
+
+type ConfigYAML struct {
+	KeyPath  string
+	CertPath string
+	DBPath   string
+	Port     int
+}
+
+type Config struct {
+	Key    []byte
+	Cert   []byte
+	DBPath string
+	Port   int
+}
+
+type Environment struct {
+	DB *certdb.CertificateRequestsRepository
+}
+
+// validateConfigFile opens and processes the given yaml file, and catches errors in the process
+func validateConfigFile(filePath string) (Config, error) {
+	validationErr := errors.New("config file validation failed: ")
+	config := Config{}
+	configYaml, err := os.ReadFile(filePath)
+	if err != nil {
+		return config, errors.Join(validationErr, err)
+	}
+	c := ConfigYAML{}
+	if err := yaml.Unmarshal(configYaml, &c); err != nil {
+		return config, errors.Join(validationErr, err)
+	}
+	cert, err := os.ReadFile(c.CertPath)
+	if err != nil {
+		return config, errors.Join(validationErr, err)
+	}
+	key, err := os.ReadFile(c.KeyPath)
+	if err != nil {
+		return config, errors.Join(validationErr, err)
+	}
+	if _, err := os.OpenFile(c.DBPath, os.O_CREATE|os.O_RDONLY, 0644); err != nil {
+		return config, errors.Join(validationErr, err)
+	}
+	config.Cert = cert
+	config.Key = key
+	config.DBPath = c.DBPath
+	config.Port = c.Port
+	return config, nil
+}
```

Comment:
Does this belong under `api`? It seems like it should be done before using api and only passing over what is needed.

Reply:
I think there could be a better place for this, but as it stands the config file is for configuring the server, and because it's very difficult to share variables between tests, it's more convenient to keep them together. If we want to move both the server and the config for the server somewhere else, I'm open to that.